src/query/storage/m3/storage.go

```diff
@@ -711,28 +711,14 @@ func (s *m3storage) Write(
 		// capture var
 		datapoint := datapoint
 		wg.Add(1)
-
-		var (
-			now                      = time.Now()
-			deadline, deadlineExists = ctx.Deadline()
-			timeout                  = minWriteWaitTimeout
-		)
-		if deadlineExists {
-			if remain := deadline.Sub(now); remain >= timeout {
-				timeout = remain
-			}
-		}
-		spawned := s.opts.WriteWorkerPool().GoWithTimeout(func() {
+		s.opts.WriteWorkerPool().AlwaysGo(func() {
 			if err := s.writeSingle(ctx, query, datapoint, id, tagIter); err != nil {
 				multiErr.add(err)
 			}
 
 			tagIter.Close()
 			wg.Done()
-		}, timeout)
-		if !spawned {
-			multiErr.add(fmt.Errorf("timeout exceeded waiting: %v", timeout))
-		}
+		})
```

Comment:
this was the bug that inspired this entire refactor. the bug is if it's not spawned, we don't call `wg.Done()` which would cause the request to hang forever. in practice that never happens, because we use a dynamic worker pool.

Reply:
wonderful.